src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java

```diff
@@ -25,76 +32,154 @@
+            this.value = value;
+            this.index = index;
+        }
+    }
 
     /**
      * (0) filter: remove unpaired reads and reads with an unmapped mate.
      * (1) keyReadsByName: label each read with its read group and read name.
      * (2) GroupByKey: group together reads with the same group and name.
-     * (3) keyPairedEndsWithAlignmentInfo:
-     *   (a) Sort each group of reads (see GATKOrder below).
-     *   (b) Pair consecutive reads into PairedEnds. In most cases there will only be two reads
-     *       with the same name. TODO: explain why there might be more.
+     * (3) keyMarkDuplicatesSparkRecords with alignment info:
+     *   (a) Generate a fragment or emptyFragment from each read if its unpaired.
+     *   (b) Pair grouped reads reads into MarkDuplicatesSparkRecord. In most cases there will only be two reads
+     *       with the same name. Mapped reads missing mates will be emitted as fragments, more than two reads will cause an exception.
      *   (c) Label each read with alignment information: Library, reference index,
      *       stranded unclipped start and reverse strand.
-     *   (d) Leftover reads are emitted, unmodified, as an unpaired end.
-     * (4) GroupByKey: Group PairedEnds that share alignment information. These pairs
+     *   (d) Unmapped Pairs, Templates of entirely non-primary reads, etc are passed through as unmarked reads
+     * (4) GroupByKey: Group MarkDuplicatesSparkRecord that share alignment information. These pairs
      *     are duplicates of each other.
      * (5) markDuplicatePairs:
      *   (a) For each group created by (4), sort the pairs by score and mark all but the
      *       highest scoring as duplicates.
      *   (b) Determine which duplicates are optical duplicates and increase the overall count.
      */
-    static JavaRDD<GATKRead> transformReads(final SAMFileHeader header, final MarkDuplicatesScoringStrategy scoringStrategy, final OpticalDuplicateFinder finder, final JavaRDD<GATKRead> reads, final int numReducers) {
+    static JavaPairRDD<IndexPair<String>, Integer> transformToDuplicateNames(final SAMFileHeader header, final MarkDuplicatesScoringStrategy scoringStrategy, final OpticalDuplicateFinder finder, final JavaRDD<GATKRead>  reads, final int numReducers) {
+        // we treat these specially and don't mark them as duplicates
+        final JavaRDD<GATKRead> mappedReads = reads.filter(ReadFilterLibrary.MAPPED::test);
+
+        final JavaPairRDD<String, Iterable<IndexPair<GATKRead>>> keyedReads = getReadsGroupedByName(header, mappedReads, numReducers);
+
+        // Place all the reads into a single RDD of MarkDuplicatesSparkRecord objects
+        final JavaPairRDD<Integer, MarkDuplicatesSparkRecord> pairedEnds = keyedReads.flatMapToPair(keyedRead -> {
+            final List<Tuple2<Integer, MarkDuplicatesSparkRecord>> out = Lists.newArrayList();
+            AtomicReference<IndexPair<GATKRead>> hadNonPrimaryRead = new AtomicReference<>();
```

Comment:
it would be good to not use an atomic reference here, they have a lot of overhead

Reply:
done